Add a more intuitive tool-tip for each add-ons rating star

SeanPrashad opened the pull request:
#Fixes #4692 - Each rating star should have its own title attribute with the rating number respective to it.

SeanPrashad commented:
I've added a `title` attribute to each button that's rendered inside `renderRatings()`. Right now, it's just a matter of figuring out the correct syntax so the template renders it correctly 😅

tofumatt commented on src/ui/components/Rating/index.js:
```diff
@@ -79,6 +79,7 @@ export class RatingBase extends React.Component {
         <button
           onClick={this.onSelectRating}
           value={thisRating}
+          title="Rate this add on %(thisRating) out of 5."
```
That's not what you'll want to do; that syntax is only for `gettext` strings. In fact, what you'll want to do here is add a `title` property to each rating above.

tofumatt replied:
(At line 58 here, you'll want to add a title attribute with `Rate this add on ${thisRating} out of 5.`... as long as the addon is not `readOnly`.)

SeanPrashad commented:
I'm trying to figure out how to access the `readOnly` variable found [here](https://github.com/mozilla/addons-frontend/blob/025b2bc03dc1049ccd9c6db2fb38f48daabf018a/src/ui/components/Rating/index.js#L53).

Ideally we need to do something like:


```js
    // ...
    key: `rating-${thisRating}`,
    ref: (ref) => { this.ratingElements[thisRating] = ref; },
    title: if (readOnly) {`Rate this add-on ${thisRating} out of 5.`}, 
  };
```

tofumatt commented on src/ui/components/Rating/index.js:
```diff
@@ -68,6 +68,7 @@ export class RatingBase extends React.Component {
         id: `Rating-rating-${thisRating}`,
         key: `rating-${thisRating}`,
         ref: (ref) => { this.ratingElements[thisRating] = ref; },
+        title: `Rate this add-on ${thisRating} out of 5.`,
```
This should be <code>title: readOnly ? \`Rate this add-on ${thisRating} out of 5.\` : null,</code>

tofumatt commented on src/ui/components/Rating/index.js:
```diff
@@ -93,11 +94,9 @@ export class RatingBase extends React.Component {
     }
     let description;
-    if (rating) {
+    if (rating && readOnly) {
```
Thanks for this! A comment here explaining why only `readOnly` rating wrappers should get this description might be good here.

SeanPrashad commented:
It looks like we're almost there but I'm trying to figure out why the `'Rated %(rating)s out of 5'` isn't rendering after I've rated an add-on.

I've installed [React Developer Tools](https://addons.mozilla.org/en-US/firefox/addon/react-devtools/) and took at look at the `title` attribute when the add-on rating has a value (and is `readOnly` at this point) and it's `undefined`. Now, I think this is due to the fact that `description` never gets initialized [here](https://github.com/mozilla/addons-frontend/pull/4693/files#diff-88b641d04bd9c7d77b5475071f0870b6R96).

This is what RDT is showing me:

<img width="828" alt="react_dev_tools_title_value" src="https://user-images.githubusercontent.com/13009507/38175434-6661187e-35aa-11e8-9336-f39df0f61910.png">

I'm going to take another look at why adding `readOnly` to our `if` statement broke things!

SeanPrashad commented:
**Update**: RDT is very useful! 👌🏽

Upon closer inspection, RDT allows us to set the values for props to mimic specific behaviour such as when an add-on receives a rating. By setting `readOnly` to true after rating an add-on, we're able to see the expected results in action:

![react_dev_tools_show_readonly](https://user-images.githubusercontent.com/13009507/38175731-15c45372-35af-11e8-8c4d-cd64c457883f.gif)

And when `readOnly` is not set to true, the tooltip behaves as expected:

![react_dev_tools_show_readonly_false](https://user-images.githubusercontent.com/13009507/38175746-4745069e-35af-11e8-873c-2e679f85e41e.gif)

Moving onto tests now!

SeanPrashad commented:
I think everyone is good to go @tofumatt! I've updated the OP with a few GIF's to showcase this patch but here they are once more:

_When the user is rating the add-on for the first time_:

![tooltip_after_update](https://user-images.githubusercontent.com/13009507/38176823-08d0e4f4-35c4-11e8-8ee4-a752b77ed8a0.gif)

_After the user has submitted a rating for the add-on_:

![tooltip_after_update_readonly](https://user-images.githubusercontent.com/13009507/38176835-4f48bee8-35c4-11e8-8325-4e9e63f12f5a.gif)

tofumatt commented on src/ui/components/Rating/index.js:
```diff
@@ -92,12 +93,13 @@ export class RatingBase extends React.Component {
     }
-    let description;
-    if (rating) {
+    let description = 'Click to rate this add-on';
```
Any strings that we show to users need to be wrapped in a gettext call; this string will not get localised for users which would be bad.

That said: why is this being set by default? Since we could be setting a `title` attribute each rating star shouldn't we skip the title on the entire rating wrapper? I don't think this is a good message anyway, it can probably just go.

tofumatt commented on src/ui/components/Rating/index.js:
```diff
@@ -68,6 +68,7 @@ export class RatingBase extends React.Component {
         id: `Rating-rating-${thisRating}`,
         key: `rating-${thisRating}`,
         ref: (ref) => { this.ratingElements[thisRating] = ref; },
+        title: !readOnly ? `Rate this add-on ${thisRating} out of 5.` : null,
```
This needs to be translated using `i18n.gettext` and `i18n.sprintf`. You can find examples on how to use them elsewhere in the app.

SeanPrashad commented:
I've gone ahead and added the respective functions to internationalize our strings! I also removed the title for the wrapper `div`. In addition, I thought it would be an appropriate time to add your suggestion for when a user is rating the add-on after their first time:

![react_dev_tools_show_postrating_postfeedback](https://user-images.githubusercontent.com/13009507/38178964-14c7b75c-35eb-11e8-9733-fb04fec261b9.gif)

Another problem though - the tooltip is shared with the smaller version of the stars, found here:

<img width="545" alt="resulting_star_bug" src="https://user-images.githubusercontent.com/13009507/38178976-529868f6-35eb-11e8-9ec0-7e7708569910.png">

What are your thoughts on this?

tofumatt commented on src/ui/components/Rating/index.js:
```diff
@@ -68,9 +68,12 @@ export class RatingBase extends React.Component {
         id: `Rating-rating-${thisRating}`,
         key: `rating-${thisRating}`,
         ref: (ref) => { this.ratingElements[thisRating] = ref; },
+        title: rating ? i18n.sprintf(i18n.gettext(`Update your rating to %(thisRating)s out of 5.`), { thisRating }) :
```
You aren’t checking for readOnly any,ore, so you’ll need to add that back. That’s why the tool tips are appearing in the small, read only versions.

SeanPrashad replied:
I moved out the check for `readOnly` to the if block that check for `readOnly` found [here](https://github.com/mozilla/addons-frontend/pull/4693/files#diff-88b641d04bd9c7d77b5475071f0870b6R76). The reason is that I could not nest ternary if statements and if I simply had `if (!readOnly && !rating) { ... }`, then tooltip would show `Update your rating to ... ` over the smaller read only stars.

This approach makes sure that the `readOnly` stars always have their `title` property removed and the clickable versions have a tooltip that dynamically reflects if the user is rating the add-on for the first or fifth time!

tofumatt replied:
For this check I'd set `rating && !readOnly`, because I don't think `readOnly` stars should each have the same duplicate title; if its `readOnly`, as mentioned below, we should just put a `title` attribute on the whole enclosing `<div>`.

SeanPrashad replied:
Thanks for the clarification - I actually removed the wrapping div as per your comment found [here](https://github.com/mozilla/addons-frontend/pull/4693#discussion_r178470860). Should we add it back and update it as necessary?

tofumatt replied:
Oh snap, I'm sorry. Yeah, I think it's worth having the wrapper `title` *only* when it's `readOnly`.

bobsilverberg commented on src/ui/components/Rating/index.js:
```diff
@@ -68,9 +68,14 @@ export class RatingBase extends React.Component {
       if (readOnly) {
+        props.title = rating ? i18n.sprintf(i18n.gettext('Rated %(rating)s out of 5.'),
+          { rating: i18n.formatNumber(parseFloat(rating).toFixed(1)) }) :
+          i18n.gettext(`Unable to retrieve rating.`);
```
Sorry for the drive-by review, but do you think a better message might be `This add-on has not been rated yet`? The `Unable` suggests (to me at least) that some sort of error occurred.

SeanPrashad replied:
Swapped out the error messages and updated the GIFs in the OP - thanks Bob!

SeanPrashad commented:
@tofumatt I'd like a review when you get a chance!

Just a quick summary of what we've done so far:

- Removed the `<div>` that wrapped all of the stars (as well as it's title)
- Updated the existing tests for the removed `<div>` to instead check the `title` of the stars
- Added 7 GIFs showcasing the various use cases that were updated

tofumatt commented:
@SeanPrashad Looks like you've got some merge conflicts with this now so you'll need to rebase your work on top of master first. After that I can take a look!

SeanPrashad commented:
☑️ Rebase complete - all ready for a review now @tofumatt!

tofumatt commented on src/ui/components/Rating/index.js:
```diff
@@ -68,9 +68,14 @@ export class RatingBase extends React.Component {
       };
 
       if (readOnly) {
+        props.title = rating ? i18n.sprintf(i18n.gettext('Rated %(rating)s out of 5.'),
```
The issue with this is it's going to set the title for each star, but if it's `readOnly` that's not needed; a `title` attribute for the entire `<div>` makes more sense.

tofumatt commented on src/ui/components/Rating/index.js:
```diff
@@ -86,19 +91,12 @@ export class RatingBase extends React.Component {
         `styleSize=${styleSize} is not a valid value; ` +
         `possible values: ${RATING_STYLE_SIZES.join(', ')}`);
     }
-    let description = i18n.gettext('This add-on has no ratings.');
```
I don't think this stuff should be removed, except for `description = i18n.gettext('Click to rate this add-on');`

That could probably be changed to set `description = null` so that the "Rate this add X out of 5" is shown for `!readOnly` add-ons.

SeanPrashad commented:
Updated as requested @tofumatt!

tofumatt commented on src/ui/components/Rating/index.js:
```diff
@@ -68,9 +68,12 @@ export class RatingBase extends React.Component {
         id: `Rating-rating-${thisRating}`,
         key: `rating-${thisRating}`,
         ref: (ref) => { this.ratingElements[thisRating] = ref; },
+        title: (rating && !readOnly) ? i18n.sprintf(i18n.gettext(`Update your rating to %(thisRating)s out of 5.`), { thisRating }) :
```
Looking at this code, if  `!rating && readOnly === true` then we will show "Rate this add-on %(thisRating)s out of 5.", which is not intended.

Did you test this out and get the desired result? This `readOnly` check seems especially superfluous given you set `props.title` to `null` later on.

tofumatt commented on src/ui/components/Rating/index.js:
```diff
@@ -68,9 +68,12 @@ export class RatingBase extends React.Component {
       };
 
       if (readOnly) {
+        props.title = null;
```
I'm not a huge fan of setting, then deleting props. I think it would be better to have an if statement that *adds* `title` to `props` if the right conditions are set.

SeanPrashad replied:
I've added another if statement to add, or update, the title of `props` if the user has previously rated the add-on. This still feels like it can be improved though so any suggestions are welcomed!

tofumatt commented on src/ui/components/Rating/index.js:
```diff
@@ -86,18 +89,16 @@ export class RatingBase extends React.Component {
+
+    let description = null;
+    if (readOnly) {
+      description = rating ? description = i18n.sprintf(i18n.gettext('Rated %(rating)s out of 5.'), { rating: i18n.formatNumber(parseFloat(rating).toFixed(1)) }) : i18n.gettext('This add-on has not been rated yet.');
```
This is a really long line; I'm surprised eslint didn't complain, but can we spread this out over a few lines anyway?

tofumatt commented on tests/unit/ui/components/TestRating.js:
```diff
@@ -180,22 +193,35 @@ describe(__filename, () => {
-    expect(findDOMNode(root).textContent).toEqual('Rated 2 out of 5');
+
+    [1, 2, 3, 4, 5].forEach((rating) => {
+      expect(root.ratingElements[rating].title).toEqual(`Update your rating to ${rating} out of 5.`);
```
I think you should be testing to make sure these are not set when `readOnly` is set and I'm not sure I see that test anywhere.

SeanPrashad replied:
I went ahead and added another test for that case, found [here](https://github.com/mozilla/addons-frontend/pull/4693/files#diff-7ac6d4cc535e144232c13e382e52eeecR224)!

tofumatt commented on tests/unit/ui/components/TestRating.js:
```diff
@@ -180,22 +193,35 @@ describe(__filename, () => {
+      expect(root.ratingElements[rating].title).toEqual(`Update your rating to ${rating} out of 5.`);
+    });
+
+    const rootReadOnly = render({ rating: 2, readOnly: true });
```
This should be part of a different (eg its own) test.

SeanPrashad commented:
All ready for another review. Here's what the use cases look like:

1. Undefined/Null Rating with read-only as true:

![undefinedrating_truereadonly](https://user-images.githubusercontent.com/13009507/38632711-87cad73c-3d8b-11e8-967a-076fa5ddb032.gif)

2. Undefined/Null Rating with read-only as false:

![undefinedrating_falsereadonly](https://user-images.githubusercontent.com/13009507/38632722-8fb7ad62-3d8b-11e8-9740-b12112d50e3b.gif)

3. Rating of 3 with read-only as true:

![3rating_truereadonly](https://user-images.githubusercontent.com/13009507/38632734-983c2c24-3d8b-11e8-8d46-0c9f47753120.gif)

4. Rating of 3 with read-only as false:

![3rating_falsereadonly](https://user-images.githubusercontent.com/13009507/38632747-9f49c706-3d8b-11e8-9b9d-7ed749517deb.gif)

5. Checking that the smaller, read-only tooltip functions as normal:

![readonlytooltip](https://user-images.githubusercontent.com/13009507/38632766-ad66ca82-3d8b-11e8-8d01-71f7ec0a12ad.gif)

SeanPrashad commented:
@caitmuenster I noticed that the original issue was closed and was wondering if I should also close this pull request?

caitmuenster commented:
@SeanPrashad That was totally an accident on my part and definitely my fault! Please keep the pull request open. :)

bobsilverberg commented on src/ui/components/Rating/index.js:
```diff
@@ -68,8 +68,13 @@ export class RatingBase extends React.Component {
         id: `Rating-rating-${thisRating}`,
         key: `rating-${thisRating}`,
         ref: (ref) => { this.ratingElements[thisRating] = ref; },
+        title: !readOnly ? i18n.sprintf(i18n.gettext(`Rate this add-on %(thisRating)s out of 5.`), { thisRating }) : null,
```
I think this `title` setting code would be a bit cleaner if you put both cases into a function, which would return the correct title, instead of setting a default title here, and then conditionally setting it again below.

Also, please try to keep the line length at a maximum of 80 characters.

SeanPrashad commented:
I've moved the logic for rendering a title into `renderTitle` and included a short comment about what the function is used for @bobsilverberg! I also rebased on master as there were some new tests added which conflicted with some of this patch.

I particularly don't like how messy the function's contents are so please let me know how I can best clean it up (if possible) 👍🏽

bobsilverberg commented on src/ui/components/Rating/index.js:
```diff
@@ -82,18 +101,18 @@ export class RatingBase extends React.Component {
+    let description = null;
+    // Wrap readOnly ratings with a description to maintain functionality
+    // for the "Average rating of developer’s add-ons" tooltip
+    if (readOnly) {
```
This could be shortened to `const description = readOnly ? this.renderTitle(i18n, rating, readOnly, null) : null;`

bobsilverberg commented on src/ui/components/Rating/index.js:
```diff
@@ -49,8 +49,26 @@ export class RatingBase extends React.Component {
+  // Helper function used to render title attributes
+  // for each individual star, as well as the wrapper
+  // that surrounds the read-only set of stars
+  renderTitle = (i18n, rating, readOnly, starRating) => {
```
This could be cleaned up a bit. For one, you could remove all the `else if` statements, because each `if` returns something, so there's no need for any `else`s.

You could also try grouping it a bit, for example:

```js
  renderTitle = (i18n, rating, readOnly, starRating) => {
    if (readOnly) {
      if (rating) {
        return i18n.sprintf(i18n.gettext('Rated %(rating)s out of 5.'),
          { rating: i18n.formatNumber(parseFloat(rating).toFixed(1)) });
      }
      return i18n.gettext('This add-on has not been rated yet.');
    }
    if (rating) {
      return i18n.sprintf(i18n.gettext(
        `Update your rating to %(starRating)s out of 5.`), { starRating });
    }
    return i18n.sprintf(i18n.gettext(
      `Rate this add-on %(starRating)s out of 5.`), { starRating });
  }
```

bobsilverberg commented on tests/unit/ui/components/TestRating.js:
```diff
@@ -142,13 +142,50 @@ describe(__filename, () => {
     expect(findDOMNode(root).title).toContain('3.6 out of 5');
   });
 
+  it('converts rating numbers to a float', () => {
```
We don't have any code that explicitly converts to float, so we don't really need this test. I believe a similar one was removed in an earlier commit.

bobsilverberg commented on tests/unit/ui/components/TestRating.js:
```diff
@@ -142,13 +142,50 @@ describe(__filename, () => {
+    const root = render({ rating: 3 });
+
+    [1, 2, 3].forEach((rating) => {
+      expect(root.ratingElements[rating].className).toEqual('Rating-choice Rating-selected-star');
```
Because this test is just for the `title`, we don't really need to assert anything about the `className` of the rating element.

bobsilverberg commented on tests/unit/ui/components/TestRating.js:
```diff
@@ -142,13 +142,50 @@ describe(__filename, () => {
+      expect(root.ratingElements[rating].title).toEqual(`Update your rating to ${rating} out of 5.`);
+    });
+    [4, 5].forEach((rating) => {
+      expect(root.ratingElements[rating].className).toEqual('Rating-choice');
```
As above, we don't really need this assertion.

bobsilverberg commented on tests/unit/ui/components/TestRating.js:
```diff
@@ -142,13 +142,50 @@ describe(__filename, () => {
+    expect(findDOMNode(rootWithString).title).toContain('3.6 out of 5');
+  });
+
+  it('rounds readOnly average ratings to nearest 0.5 multiple', () => {
```
This test was replaced with two others above it, specifically the ones for `it('renders correct full stars for a rating')` and `it('renders correct half stars for a rating')`, so this test can be removed.

bobsilverberg commented on src/ui/components/Rating/index.js:
```diff
@@ -49,8 +49,27 @@ export class RatingBase extends React.Component {
+
   renderRatings() {
-    const { readOnly } = this.props;
+    const { i18n, readOnly } = this.props;
```
Oops, sorry I missed this last time, but I don't think you need to accept extract `i18n` from `this.props` here, only to pass it into `this.renderTitle`. You could just add a `const { i18n } = this.props;` at the top of `renderTitle`, I believe.

SeanPrashad replied:
I went ahead and implemented it but it broke about 8 tests (which is probably too large to paste here). Should I go ahead and refactor them?

willdurand commented:
Hi @SeanPrashad, is it ready for another look? thanks!

SeanPrashad commented:
Hey @willdurand! Yes it should be ready to go - @bobsilverberg and I were able to polish up one final thing recently.

willdurand commented on src/ui/components/Rating/index.js:
```diff
@@ -49,6 +49,27 @@ export class RatingBase extends React.Component {
 
+  // Helper function used to render title attributes
+  // for each individual star, as well as the wrapper
+  // that surrounds the read-only set of stars
```
Nitpick: missing period.

willdurand commented on src/ui/components/Rating/index.js:
```diff
@@ -49,6 +49,27 @@ export class RatingBase extends React.Component {
+      return i18n.sprintf(i18n.gettext(
+        `Update your rating to %(starRating)s out of 5.`), { starRating });
+    }
+    return i18n.sprintf(i18n.gettext(
```
could you add a blank line before this `return`?

willdurand commented on src/ui/components/Rating/index.js:
```diff
@@ -49,6 +49,27 @@ export class RatingBase extends React.Component {
+      }
+      return i18n.gettext('This add-on has not been rated yet.');
+    }
+    if (rating) {
```
could you add a blank line before this `if`?

willdurand commented on src/ui/components/Rating/index.js:
```diff
@@ -82,19 +104,16 @@ export class RatingBase extends React.Component {
-    }
+
+    // Wrap read only ratings with a description to maintain functionality
+    // for the "Average rating of developer’s add-ons" tooltip
```
Nitpick again: missing period.

willdurand commented on tests/unit/ui/components/TestRating.js:
```diff
@@ -142,9 +142,20 @@ describe(__filename, () => {
+      expect(root.ratingElements[rating].title).toEqual(`Update your rating to ${rating} out of 5.`);
+    });
+    [4, 5].forEach((rating) => {
+      expect(root.ratingElements[rating].title).toEqual(`Update your rating to ${rating} out of 5.`);
```
I don't understand why there are two arrays here.

tofumatt replied:
That is a holderover from the old test I think, but if the rating is 3 this should be updated to have "Update your rating to X" for everything except 3, and then a "Your rating is currently 3/5" or even just no title for 3...

SeanPrashad replied:
@tofumatt is right - here's the old test for reference: https://github.com/mozilla/addons-frontend/blob/master/tests/unit/ui/components/TestRating.js#L195

willdurand commented on tests/unit/ui/components/TestRating.js:
```diff
@@ -171,22 +182,40 @@ describe(__filename, () => {
   });
 
-  it('renders "no ratings" if no rating and readOnly', () => {
+  it('renders an accessible description for null ratings and readOnly', () => {
```
s/readOnly/read only (or read-only)

SeanPrashad commented:
I've addressed the nits and moved the read-only tests into the read-only section. I noticed that there was one tests that was repeated so I removed that one.

willdurand commented on tests/unit/ui/components/TestRating.js:
```diff
@@ -253,22 +244,66 @@ describe(__filename, () => {
 
     it('renders empty ratings', () => {
-      expect(getRating({ rating: null })).toEqual('Click to rate this add-on');
+      expect(getRating({ rating: null, readOnly: true })).toEqual('This add-on has not been rated yet.');
```
You should try to wrap lines at max 80 chars.

SeanPrashad replied:
Thanks for that - I am wondering why linting isn't giving me an error.. locally or via Travis.

willdurand replied:
I dont think we have a rule for that.

SeanPrashad replied:
It might be worthwhile to use [Prettier pre-commit hooks](https://prettier.io/docs/en/precommit.html) to automatically format code  when commits are made based on the existing `eslint` config.

bobsilverberg commented:
Thanks for all of your hard work on this @SeanPrashad. It looks like @willdurand took a final look and gave it his seal of approval, so I'm going to merge it.

TonnesM commented:
Please remember to not use trailing periods for action tooltips, as they make a string instructive/imperative rather than infinitive, which will reflect in a number of localizations, but may also affect the meaning for en-US. (Localizers sometimes fully depend on periods being present or not in order to see what’s intended rather than testing them.)

I.e. they would read

"Rate this add-on / Update your rating to %(thisRating)s out of 5!"

instead of 

(User can or wants to) "Rate this add-on / Update their rating to %(thisRating)s out of 5" (by clicking the star)

A change/fix in any next push would be welcome.

bobsilverberg commented:
@SeanPrashad Do you want to open a follow-up issue for the change suggested above?

SeanPrashad commented:
Yes! I'll open another PR when I get home from work later today.
